andela/apollo-ah-backend
# #164521891 Implements reset password

**What does this PR do?**
This PR implements reset password feature

**Description of Task to be completed**

- [x] Write unit and integration tests
- [x] Create api endpoint PATCH /api/v1/users/reset_password to handle reset password requests
- [x] Create and mount middleware to validate password parameters and authentication token
- [x] Update usersController class to with function that updates the user password


#### How should this be manually tested

_Prerequisite: Postgres_

1. Run the following commands on terminal:

```bash
git clone -b feature/164521891/reset-password --change-branch \
https://github.com/andela/apollo-ah-backend.git

Navigate to the apollo-ah-backend directory
Run cp .env.example .env
npm install
```

2. Modify `.env` file with your db credentials

3. Run migrations with `npm run migrate`

4. Spin up the server by running `npm run start:dev` on your terminal

5. Make a POST request to `localhost:3000/api/v1/users/signup` with [Postman](https://www.getpostman.com/downloads/). to create an account

6. Make a POST request to `localhost:3000/api/v1/users/forgot_password` with [Postman](https://www.getpostman.com/downloads/) using the email from the account created to request for a new password. Check your email and copy the token parameter from the link provided.

7. Make a POST request to `localhost:3000/api/v1/users/reset_password?token={TOKEN}` with [Postman](https://www.getpostman.com/downloads/) where TOKEN is the token from step 6 above.

❗️ **Important:** Provide user payload when creating an account, sample below:
```json
{
  "email": "example@email.com",
  "password": "secret"
  "username" "user"
}
```
For forgot password endpoint, use the same email from above to make request using this format,
```json
{
  "email": "example@email.com",
}
```
For reset password endpoint, use token gotten from the previous step,
```json
{
  "password": "new password",
  "confirmPassword": "new password"
}
```

**NOTE:** To run test locally, create `.env.test` file. Run command: `cp .env.example .env.test`

#### Any background context you want to provide
N/A

#### What are the relevant pivotal tracker stories

[#164521891](https://www.pivotaltracker.com/story/show/164521891)

#### Screenshots
N/A

_-- Template format by @devdbrandy_
